File: snippets/csharp/System/NullReferenceException/Overview/Chain1.cs
using System;

namespace Chain1
{
    // <Snippet6>
    public class Chain1Example
    {
        public static void Main()
        {
            var pages = new Pages();
            if (!string.IsNullOrEmpty(pages.CurrentPage.Title))
            {
                string title = pages.CurrentPage.Title;
                Console.WriteLine($"Current title: '{title}'");
            }
        }
    }

    public class Pages
    {
        readonly Page[] _page = new Page[10];
        int _ctr = 0;

        public Page CurrentPage
        {
            get { return _page[_ctr]; }
            set
            {
                // Move all the page objects down to accommodate the new one.
                if (_ctr > _page.GetUpperBound(0))
                {
                    for (int ndx = 1; ndx <= _page.GetUpperBound(0); ndx++)
                        _page[ndx - 1] = _page[ndx];
                }
                _page[_ctr] = value;
                if (_ctr < _page.GetUpperBound(0))
                    _ctr++;
            }
        }

        public Page PreviousPage
        {
            get
            {
                if (_ctr == 0)
                {
                    if (_page[0] is null)
                        return null;
                    else
                        return _page[0];
                }
                else
                {
                    _ctr--;
                    return _page[_ctr + 1];
                }
            }
        }
    }

    public class Page
    {
        public Uri URL;
        public string Title;
    }

    // The example displays the following output:
    //    Unhandled Exception:
    //       System.NullReferenceException: Object reference not set to an instance of an object.
    //       at Chain1Example.Main()
    // </Snippet6>
}

File: snippets/fsharp/System/NullReferenceException/Overview/Chain1.fs
module Chain1

// <Snippet6>
open System

type Page() =
    [<DefaultValue>]
    val mutable public URL: Uri
    [<DefaultValue>]
    val mutable public Title: string

type Pages() =
    let pages = Array.zeroCreate<Page> 10
    let mutable i = 0

    member _.CurrentPage
        with get () = pages[i]
        and set (value) =
            // Move all the page objects down to accommodate the new one.
            if i > pages.GetUpperBound 0 then
                for ndx = 1 to pages.GetUpperBound 0 do
                    pages[ndx - 1] <- pages[ndx]

            pages[i] <- value
            if i < pages.GetUpperBound 0 then
                i <- i + 1

    member _.PreviousPage =
        if i = 0 then
            if box pages[0] = null then
                Unchecked.defaultof<Page>
            else
                pages[0]
        else
            i <- i - 1
            pages[i + 1]

let pages = Pages()
if String.IsNullOrEmpty pages.CurrentPage.Title |> not then
    let title = pages.CurrentPage.Title
    printfn $"Current title: '{title}'"


// The example displays the following output:
//    Unhandled Exception:
//       System.NullReferenceException: Object reference not set to an instance of an object.
//       at <StartupCode$fs>.main()
// </Snippet6>